File: backend/app/schemas/diary.py
"""情绪日记相关请求/响应模型。

包含日记 CRUD、隐私同意、同步设置、导出等接口的 Schema 定义。
"""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Any

from pydantic import Field, field_validator, model_validator

from app.schemas.base import BaseSchema


# ---------------------------------------------------------------------------
# 情绪色调枚举
# ---------------------------------------------------------------------------

class EmotionTone(str, Enum):
    """情绪色调枚举。

    五种色调代表不同情绪能量级别：
    - warm_orange: 充满能量、开心
    - light_green: 平静、安稳
    - gray_blue: 低落、沉闷
    - deep_blue: 难过、忧伤
    - dark_purple: 崩溃、混乱
    """

    WARM_ORANGE = "warm_orange"  # 暖橘 #FF9A5C
    LIGHT_GREEN = "light_green"  # 浅绿 #8FCCA0
    GRAY_BLUE = "gray_blue"  # 灰蓝 #8BA7C4
    DEEP_BLUE = "deep_blue"  # 深蓝 #4A6FA5
    DARK_PURPLE = "dark_purple"  # 暗紫 #6B4C7A


# 情绪色调映射：色调 -> 可选标签列表
EMOTION_TONE_LABELS: dict[str, list[str]] = {
    EmotionTone.WARM_ORANGE.value: [
        "开心", "感恩", "兴奋", "被爱", "有希望", "自豪", "释然",
    ],
    EmotionTone.LIGHT_GREEN.value: [
        "平静", "放松", "专注", "安心", "满足", "无聊",
    ],
    EmotionTone.GRAY_BLUE.value: [
        "焦虑", "疲惫", "迷茫", "孤独", "委屈", "烦躁",
    ],
    EmotionTone.DEEP_BLUE.value: [
        "难过", "失望", "自责", "心疼", "想念", "害怕",
    ],
    EmotionTone.DARK_PURPLE.value: [
        "混乱", "麻木", "空洞", "矛盾", "崩溃", "说不清",
    ],
}

# 色调元数据：颜色代码和代表语
EMOTION_TONE_META: dict[str, dict[str, str]] = {
    EmotionTone.WARM_ORANGE.value: {
        "color": "#FF9A5C",
        "meaning": "充满能量、开心",
        "phrase": "今天还不错",
    },
    EmotionTone.LIGHT_GREEN.value: {
        "color": "#8FCCA0",
        "meaning": "平静、安稳",
        "phrase": "还算正常",
    },
    EmotionTone.GRAY_BLUE.value: {
        "color": "#8BA7C4",
        "meaning": "低落、沉闷",
        "phrase": "有点堵",
    },
    EmotionTone.DEEP_BLUE.value: {
        "color": "#4A6FA5",
        "meaning": "难过、忧伤",
        "phrase": "很难受",
    },
    EmotionTone.DARK_PURPLE.value: {
        "color": "#6B4C7A",
        "meaning": "崩溃、混乱",
        "phrase": "说不清",
    },
}


# ---------------------------------------------------------------------------
# 同步模式枚举
# ---------------------------------------------------------------------------

class SyncMode(str, Enum):
    """日记同步模式枚举。

    - LOCAL_ONLY: 仅存储在本地设备
    - CLOUD_SYNC: 开启云端同步
    """

    LOCAL_ONLY = "local_only"
    CLOUD_SYNC = "cloud_sync"


# ---------------------------------------------------------------------------
# 导出格式枚举
# ---------------------------------------------------------------------------

class ExportFormat(str, Enum):
    """日记导出格式枚举。"""

    JSON = "json"
    PDF = "pdf"


# ---------------------------------------------------------------------------
# 日记创建/更新请求
# ---------------------------------------------------------------------------

class DiaryCreateRequest(BaseSchema):
    """创建日记请求模型。

    三层标签结构：
    - emotion_tone: 情绪色调（必选1）
    - emotion_labels: 情绪标签（可选，最多3个）
    - content_text: 自由文字（可选，支持语音输入）
    """

    emotion_tone: EmotionTone = Field(
        ...,
        description="情绪色调",
        examples=["warm_orange"],
    )
    emotion_labels: list[str] | None = Field(
        default=None,
        max_length=3,
        description="情绪标签列表，最多3个",
        examples=[["开心", "感恩"]],
    )
    content_text: str | None = Field(
        default=None,
        max_length=2000,
        description="日记内容文字，最多2000字",
        examples=["今天天气真好，心情愉快。"],
    )
    record_date: date = Field(
        ...,
        description="记录日期",
        examples=["2024-01-15"],
    )
    client_id: str | None = Field(
        default=None,
        max_length=50,
        description="客户端唯一标识，用于离线同步去重",
        examples=["device_abc123"],
    )
    is_encrypted: bool = Field(
        default=False,
        description="内容是否已加密",
    )
    content_hash: str | None = Field(
        default=None,
        max_length=64,
        description="内容哈希，用于完整性校验",
    )

    @field_validator("emotion_labels", mode="before")
    @classmethod
    def validate_emotion_labels(cls, v: list[str] | None) -> list[str] | None:
        """验证情绪标签格式。"""
        if v is None:
            return None
        # 去重并去空
        labels = [label.strip() for label in v if label and label.strip()]
        if len(labels) == 0:
            return None
        return labels[:3]  # 最多3个

    @model_validator(mode="after")
    def validate_labels_match_tone(self) -> "DiaryCreateRequest":
        """验证标签是否与色调匹配。"""
        if self.emotion_labels:
            valid_labels = EMOTION_TONE_LABELS.get(self.emotion_tone.value, [])
            for label in self.emotion_labels:
                if label not in valid_labels:
                    raise ValueError(
                        f"情绪标签 '{label}' 不属于色调 '{self.emotion_tone.value}' 的有效标签"
                    )
        return self


class DiaryUpdateRequest(BaseSchema):
    """更新日记请求模型。"""

    emotion_tone: EmotionTone | None = Field(
        default=None,
        description="情绪色调",
    )
    emotion_labels: list[str] | None = Field(
        default=None,
        max_length=3,
        description="情绪标签列表，最多3个",
    )
    content_text: str | None = Field(
        default=None,
        max_length=2000,
        description="日记内容文字",
    )
    is_encrypted: bool = Field(
        default=False,
        description="内容是否已加密",
    )
    content_hash: str | None = Field(
        default=None,
        max_length=64,
        description="内容哈希，用于完整性校验",
    )

    @field_validator("emotion_labels", mode="before")
    @classmethod
    def validate_emotion_labels(cls, v: list[str] | None) -> list[str] | None:
        """验证情绪标签格式。"""
        if v is None:
            return None
        labels = [label.strip() for label in v if label and label.strip()]
        if len(labels) == 0:
            return None
        return labels[:3]

    @model_validator(mode="after")
    def validate_labels_match_tone(self) -> "DiaryUpdateRequest":
        """验证标签是否与色调匹配。"""
        if self.emotion_tone and self.emotion_labels:
            valid_labels = EMOTION_TONE_LABELS.get(self.emotion_tone.value, [])
            for label in self.emotion_labels:
                if label not in valid_labels:
                    raise ValueError(
                        f"情绪标签 '{label}' 不属于色调 '{self.emotion_tone.value}' 的有效标签"
                    )
        return self


# ---------------------------------------------------------------------------
# 日记响应模型
# ---------------------------------------------------------------------------

class DiaryResponse(BaseSchema):
    """日记响应模型。"""

    id: str = Field(..., description="日记ID")
    emotion_tone: str | None = Field(None, description="情绪色调")
    emotion_labels: list[str] | None = Field(None, description="情绪标签列表")
    content_text: str | None = Field(None, description="日记内容（已解密）")
    record_date: date = Field(..., description="记录日期")
    is_synced: bool = Field(..., description="是否已同步")
    is_encrypted: bool = Field(default=False, description="内容是否加密")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    # 计算属性：是否为 0 字记录
    is_zero_record: bool = Field(
        default=False,
        description="是否为0字记录（纯色调，不计入周报分析）",
    )


class DiaryListResponse(BaseSchema):
    """日记列表响应模型。"""

    data: list[DiaryResponse] = Field(default_factory=list, description="日记列表")
    pagination: dict[str, Any] = Field(..., description="分页信息")


class DiaryDetailResponse(BaseSchema):
    """日记详情响应模型。"""

    id: str = Field(..., description="日记ID")
    emotion_tone: str | None = Field(None, description="情绪色调")
    emotion_labels: list[str] | None = Field(None, description="情绪标签列表")
    content_text: str | None = Field(None, description="日记内容（已解密）")
    record_date: date = Field(..., description="记录日期")
    is_synced: bool = Field(..., description="是否已同步")
    is_encrypted: bool = Field(default=False, description="内容是否加密")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    client_id: str | None = Field(None, description="客户端ID")
    content_hash: str | None = Field(None, description="内容哈希")

    # 计算属性
    is_zero_record: bool = Field(
        default=False,
        description="是否为0字记录",
    )
    tone_meta: dict[str, str] | None = Field(
        None,
        description="色调元数据（颜色、含义、代表语）",
    )


# ---------------------------------------------------------------------------
# 隐私同意相关
# ---------------------------------------------------------------------------

class PrivacyConsentResponse(BaseSchema):
    """隐私同意状态响应模型。"""

    has_consented: bool = Field(..., description="是否已同意隐私声明")
    consented_at: datetime | None = Field(None, description="同意时间")
    sync_mode: SyncMode = Field(
        default=SyncMode.LOCAL_ONLY,
        description="当前同步模式",
    )


class PrivacyConsentRequest(BaseSchema):
    """同意隐私声明请求模型。"""

    sync_mode: SyncMode = Field(
        ...,
        description="选择的同步模式：local_only 或 cloud_sync",
    )


# ---------------------------------------------------------------------------
# 同步设置相关
# ---------------------------------------------------------------------------

class SyncSettingsResponse(BaseSchema):
    """同步设置响应模型。"""

    sync_mode: SyncMode = Field(..., description="当前同步模式")
    last_sync_at: datetime | None = Field(None, description="上次同步时间")
    sync_device_count: int = Field(default=0, description="已同步设备数")
    encryption_enabled: bool = Field(
        default=True,
        description="是否启用端到端加密",
    )


class SyncSettingsUpdateRequest(BaseSchema):
    """更新同步设置请求模型。"""

    sync_mode: SyncMode = Field(..., description="同步模式")


# ---------------------------------------------------------------------------
# 导出相关
# ---------------------------------------------------------------------------

class ExportRequest(BaseSchema):
    """日记导出请求模型。"""

    format: ExportFormat = Field(
        default=ExportFormat.JSON,
        description="导出格式：json 或 pdf",
    )
    start_date: date | None = Field(
        default=None,
        description="导出起始日期",
    )
    end_date: date | None = Field(
        default=None,
        description="导出结束日期",
    )
    include_encrypted: bool = Field(
        default=False,
        description="是否包含加密内容（需提供密钥）",
    )

    @model_validator(mode="after")
    def validate_date_range(self) -> "ExportRequest":
        """验证日期范围。"""
        if self.start_date and self.end_date:
            if self.start_date > self.end_date:
                raise ValueError("起始日期不能晚于结束日期")
        return self


class ExportResponse(BaseSchema):
    """日记导出响应模型。"""

    download_url: str = Field(..., description="下载链接")
    expires_at: datetime = Field(..., description="链接过期时间")
    file_format: str = Field(..., description="文件格式")
    record_count: int = Field(..., description="导出记录数")


# ---------------------------------------------------------------------------
# 删除相关
# ---------------------------------------------------------------------------

class DeleteAllResponse(BaseSchema):
    """删除全部日记响应模型。"""

    deleted_count: int = Field(..., description="已删除数量")
    message: str = Field(default="所有日记已删除", description="提示信息")


# ---------------------------------------------------------------------------
# 统计相关
# ---------------------------------------------------------------------------

class DiaryStatsResponse(BaseSchema):
    """日记统计响应模型。"""

    total_records: int = Field(..., description="总记录数")
    total_days: int = Field(..., description="已记录天数")
    zero_record_count: int = Field(..., description="0字记录数")
    valid_sample_count: int = Field(..., description="有效样本数（用于周报分析）")
    emotion_distribution: dict[str, int] = Field(
        default_factory=dict,
        description="情绪分布统计",
    )

File: frontend/src/api/diary.ts
/**
 * 回声 - 情绪日记接口
 * 文件：src/api/diary.ts
 * 说明：情绪日记相关接口，包括创建日记、隐私同意、同步设置等
 */

import api from './index'
import type { ApiResponse } from '../types'

// ==================== 类型定义 ====================

/** 情绪色调类型 */
export type EmotionTone = 'warm_orange' | 'light_green' | 'gray_blue' | 'deep_blue' | 'dark_purple'

/** 同步模式类型 */
export type SyncMode = 'local_only' | 'cloud_sync'

/** 情绪色调元数据 */
export interface EmotionToneMeta {
  /** 色调值 */
  tone: EmotionTone
  /** 颜色代码 */
  color: string
  /** 含义描述 */
  meaning: string
  /** 代表语 */
  phrase: string
  /** 提示语 */
  hint: string
}

/** 创建日记参数 */
export interface CreateDiaryParams {
  /** 情绪色调 */
  emotion_tone: EmotionTone
  /** 情绪标签列表，最多3个 */
  emotion_labels?: string[]
  /** 日记内容文字，最多2000字 */
  content_text?: string
  /** 记录日期 YYYY-MM-DD */
  record_date: string
  /** 客户端唯一标识 */
  client_id?: string
  /** 内容是否已加密 */
  is_encrypted?: boolean
}

/** 日记响应 */
export interface DiaryResponse {
  /** 日记ID */
  id: string
  /** 情绪色调 */
  emotion_tone: EmotionTone | null
  /** 情绪标签列表 */
  emotion_labels: string[] | null
  /** 日记内容 */
  content_text: string | null
  /** 记录日期 */
  record_date: string
  /** 是否已同步 */
  is_synced: boolean
  /** 是否加密 */
  is_encrypted: boolean
  /** 创建时间 */
  created_at: string
  /** 更新时间 */
  updated_at: string
  /** 是否为0字记录 */
  is_zero_record: boolean
}

/** 隐私同意状态响应 */
export interface PrivacyConsentResponse {
  /** 是否已同意隐私声明 */
  has_consented: boolean
  /** 同意时间 */
  consented_at: string | null
  /** 当前同步模式 */
  sync_mode: SyncMode
}

/** 设置隐私同意参数 */
export interface PrivacyConsentRequest {
  /** 同步模式 */
  sync_mode: SyncMode
}

// ==================== 情绪色调常量 ====================

/** 情绪色调元数据映射 */
export const EMOTION_TONE_META: Record<EmotionTone, EmotionToneMeta> = {
  warm_orange: {
    tone: 'warm_orange',
    color: '#FF9A5C',
    meaning: '充满能量、开心',
    phrase: '今天还不错',
    hint: '今天什么好事发生了？',
  },
  light_green: {
    tone: 'light_green',
    color: '#8FCCA0',
    meaning: '平静、安稳',
    phrase: '还算正常',
    hint: '随便聊聊今天吧',
  },
  gray_blue: {
    tone: 'gray_blue',
    color: '#8BA7C4',
    meaning: '低落、沉闷',
    phrase: '有点堵',
    hint: '什么事让你觉得堵？',
  },
  deep_blue: {
    tone: 'deep_blue',
    color: '#4A6FA5',
    meaning: '难过、忧伤',
    phrase: '很难受',
    hint: '如果不想说具体的，说说那种感觉也行',
  },
  dark_purple: {
    tone: 'dark_purple',
    color: '#6B4C7A',
    meaning: '崩溃、混乱',
    phrase: '说不清',
    hint: '这里只有你自己，写什么都行',
  },
}

/** 情绪色调列表（按顺序） */
export const EMOTION_TONE_LIST: EmotionTone[] = [
  'warm_orange',
  'light_green',
  'gray_blue',
  'deep_blue',
  'dark_purple',
]

/** 情绪标签池 */
export const EMOTION_LABELS_POOL: Record<EmotionTone, string[]> = {
  warm_orange: ['开心', '感恩', '兴奋', '被爱', '有希望', '自豪', '释然'],
  light_green: ['平静', '放松', '专注', '安心', '满足', '无聊'],
  gray_blue: ['焦虑', '疲惫', '迷茫', '孤独', '委屈', '烦躁'],
  deep_blue: ['难过', '失望', '自责', '心疼', '想念', '害怕'],
  dark_purple: ['混乱', '麻木', '空洞', '矛盾', '崩溃', '说不清'],
}

// ==================== API 接口 ====================

/**
 * 创建日记
 * @param params 创建参数
 * @returns 日记响应
 */
export function createDiary(params: CreateDiaryParams): Promise<DiaryResponse> {
  return api.post<DiaryResponse>('/diaries', params)
}

/**
 * 获取隐私同意状态
 * @returns 隐私同意状态
 */
export function getPrivacyConsent(): Promise<PrivacyConsentResponse> {
  return api.get<PrivacyConsentResponse>('/diaries/privacy')
}

/**
 * 设置隐私同意
 * @param params 同意参数
 * @returns 隐私同意状态
 */
export function setPrivacyConsent(params: PrivacyConsentRequest): Promise<PrivacyConsentResponse> {
  return api.post<PrivacyConsentResponse>('/diaries/privacy', params)
}

/**
 * 获取日记详情
 * @param id 日记ID
 * @returns 日记详情
 */
export function getDiaryDetail(id: string): Promise<DiaryResponse> {
  return api.get<DiaryResponse>(`/diaries/${id}`)
}

/**
 * 更新日记
 * @param id 日记ID
 * @param params 更新参数
 * @returns 日记响应
 */
export function updateDiary(
  id: string,
  params: Partial<CreateDiaryParams>
): Promise<DiaryResponse> {
  return api.put<DiaryResponse>(`/diaries/${id}`, params)
}

/**
 * 删除日记
 * @param id 日记ID
 */
export function deleteDiary(id: string): Promise<void> {
  return api.delete(`/diaries/${id}`)
}

/**
 * 获取日记列表
 * @param params 查询参数
 * @returns 日记列表
 */
export function getDiaryList(params?: {
  /** 起始日期 */
  start_date?: string
  /** 结束日期 */
  end_date?: string
  /** 页码 */
  page?: number
  /** 每页数量 */
  page_size?: number
}): Promise<{
  data: DiaryResponse[]
  pagination: {
    page: number
    page_size: number
    total: number
    total_pages: number
  }
}> {
  return api.get('/diaries', params)
}

/**
 * 获取日记统计
 * @returns 统计数据
 */
export function getDiaryStats(): Promise<{
  total_records: number
  total_days: number
  zero_record_count: number
  valid_sample_count: number
  emotion_distribution: Record<string, number>
}> {
  return api.get('/diaries/stats')
}

/**
 * 删除全部日记
 */
export function deleteAllDiaries(): Promise<void> {
  return api.delete('/diaries/all')
}

/**
 * 导出日记
 * @param format 导出格式 json/pdf
 * @param params 导出参数
 * @returns 导出文件URL或数据
 */
export function exportDiaries(
  format: 'json' | 'pdf',
  params?: {
    start_date?: string
    end_date?: string
  }
): Promise<{ file_url: string; file_name: string }> {
  return api.post('/diaries/export', { format, ...params })
}

// ==================== 情绪周报相关 ====================

/** 情绪周报响应 */
export interface WeeklyReportResponse {
  /** 周报ID */
  id: string
  /** 本周起始日期（周一） */
  week_start_date: string
  /** 本周结束日期（周日） */
  week_end_date: string
  /** 动态标题，如'这周像一场漫长的周三' */
  title: string | null
  /** 情绪故事线，叙事体描述本周情绪走势 */
  story_line: string | null
  /** 情绪关键词列表 */
  keywords: string[] | null
  /** 一句看见，提炼核心感受 */
  insight: string | null
  /** 温和建议，措辞谨慎 */
  suggestion: string | null
  /** 下周展望，一句话收束 */
  outlook: string | null
  /** 本周分析日记数量 */
  diary_count: number
  /** 生成时间 */
  created_at: string
  /** 是否为空周报（本周无有效日记） */
  is_empty: boolean
  /** 是否来自缓存 */
  is_cached: boolean
}

/** 空周报响应 */
export interface EmptyWeeklyReportResponse {
  /** 本周起始日期（周一） */
  week_start_date: string
  /** 本周结束日期（周日） */
  week_end_date: string
  /** 是否为空周报 */
  is_empty: true
  /** 提示信息 */
  message: string
  /** 本周有效日记数量 */
  diary_count: number
}

/** 周报历史项 */
export interface WeeklyReportHistoryItem {
  /** 周报ID */
  id: string
  /** 本周起始日期（周一） */
  week_start_date: string
  /** 本周结束日期（周日） */
  week_end_date: string
  /** 动态标题 */
  title: string | null
  /** 一句看见 */
  insight: string | null
  /** 本周分析日记数量 */
  diary_count: number
  /** 生成时间 */
  created_at: string
}

/** 周报历史响应 */
export interface WeeklyReportHistoryResponse {
  /** 周报列表 */
  data: WeeklyReportHistoryItem[]
  /** 分页信息 */
  pagination: {
    page: number
    page_size: number
    total: number
    total_pages: number
  }
}

/**
 * 获取本周情绪周报
 * @param forceRefresh 是否强制重新生成
 * @returns 周报响应
 */
export function getWeeklyReport(
  forceRefresh = false
): Promise<WeeklyReportResponse | EmptyWeeklyReportResponse> {
  return api.get('/diaries/report/weekly', { force_refresh: forceRefresh })
}

/**
 * 获取周报历史
 * @param params 分页参数
 * @returns 周报历史列表
 */
export function getWeeklyReportHistory(params?: {
  /** 页码 */
  page?: number
  /** 每页数量 */
  page_size?: number
}): Promise<WeeklyReportHistoryResponse> {
  return api.get('/diaries/report/history', params)
}

/**
 * 类型守卫：判断是否为空周报响应
 * @param response 周报响应
 * @returns 是否为空周报
 */
export function isEmptyReport(
  response: WeeklyReportResponse | EmptyWeeklyReportResponse
): response is EmptyWeeklyReportResponse {
  return 'is_empty' in response && response.is_empty === true
}
